provider/vsphere/environ_availzones.go

```diff
@@ -63,37 +43,83 @@ func (env *environ) AvailabilityZones(ctx context.ProviderCallContext) (zones []
 
 // AvailabilityZones is part of the common.ZonedEnviron interface.
 func (env *sessionEnviron) AvailabilityZones(ctx context.ProviderCallContext) ([]common.AvailabilityZone, error) {
-	if env.zones == nil {
-		computeResources, err := env.client.ComputeResources(env.ctx)
+	if len(env.zones) > 0 {
+		// This is relatively expensive to compute, so cache it on the session
+		return env.zones, nil
+	}
+
+	folders, err := env.client.Folders(env.ctx)
+	if err != nil {
+		HandleCredentialError(err, env, ctx)
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("host folder InventoryPath=%q, Name=%q",
+		folders.HostFolder.InventoryPath, folders.HostFolder.Name())
+	hostFolder := folders.HostFolder.InventoryPath
+
+	computeResources, err := env.client.ComputeResources(env.ctx)
+	if err != nil {
+		HandleCredentialError(err, env, ctx)
+		return nil, errors.Trace(err)
+	}
+	var zones []common.AvailabilityZone
+	for _, cr := range computeResources {
+		if cr.Resource.Summary.GetComputeResourceSummary().EffectiveCpu == 0 {
+			logger.Debugf("skipping empty compute resource %q", cr.Resource.Name)
+			continue
+		}
+
+		// Add an availability zone for each resource pool under this compute
+		// resource, eg: "/DataCenter1/host/Host1/Resources"
+		pools, err := env.client.ResourcePools(env.ctx, cr.Path+"/...")
 		if err != nil {
 			HandleCredentialError(err, env, ctx)
 			return nil, errors.Trace(err)
 		}
-		var zones []common.AvailabilityZone
-		for _, cr := range computeResources {
-			if cr.Summary.GetComputeResourceSummary().EffectiveCpu == 0 {
-				logger.Debugf("skipping empty compute resource %q", cr.Name)
-				continue
-			}
-			pools, err := env.client.ResourcePools(env.ctx, cr.Name+"/...")
-			if err != nil {
-				HandleCredentialError(err, env, ctx)
-				return nil, errors.Trace(err)
-			}
-			for _, pool := range pools {
-				zone := &vmwareAvailZone{
-					r:    *cr,
-					pool: pool,
-				}
-				logger.Tracef("zone: %q", zone.Name())
-				zones = append(zones, zone)
+		for _, pool := range pools {
+			zone := &vmwareAvailZone{
+				r:    *cr.Resource,
+				pool: pool,
+				name: makeAvailZoneName(hostFolder, cr.Path, pool.InventoryPath),
 			}
+			logger.Tracef("zone: %s (cr.Name=%q pool.InventoryPath=%q)",
+				zone.Name(), zone.r.Name, zone.pool.InventoryPath)
+			zones = append(zones, zone)
+		}
+	}
+
+	if logger.IsDebugEnabled() {
+		zoneNames := make([]string, len(zones))
+		for i, zone := range zones {
+			zoneNames[i] = zone.Name()
 		}
-		env.zones = zones
+		logger.Debugf("fetched availability zones: %q", zoneNames)
 	}
+
+	env.zones = zones
 	return env.zones, nil
 }
 
+// makeAvailZoneName constructs a Vsphere availability zone name from the
+// given paths. Basically it's the path relative to the host folder without
+// the extra "Resources" path segment (which doesn't appear in the UI).
```

Comment:
I wonder if it is worth moving some of the examples that were inline into the documentation for the function.